### PR TITLE
Txt record quoting and escaping

### LINKED ISSUE #58: TXT record quoting treated inconsistently across providers
quoting of TXT records is not being applied consistently across all providers

1. (unquoted) text without whitespace appears to be applied correctly across all providers
2. quoted text appears to be applied correctly across all providers
3. (unquoted) text with whitespace is not applied correctly on GoogleCloudDNS *BUG*
4. diff incorrectly detects a difference for unquoted text on GoogleCloudDNS *BUG*
5. diff incorrectly detects a difference for quoted text on DynECT *BUG*

given the following TXT record definitions in yml...

``` yaml
  - type: TXT
    fqdn: txt.shopify.test.
    ttl: 30
    txtdata: text

  - type: TXT
    fqdn: txt.shopify.test.
    ttl: 30
    txtdata: "text with whitespace"

  - type: TXT
    fqdn: txt.shopify.test.
    ttl: 30
    txtdata: '"quoted text with whitespace"'
```

after applying in all three providers (DNSimple, DynECT, and GoogleCloudDNS), `dig` shows the following records...

```
$ dig +short TXT txt.shopify.test @ns1.dnsimple.com
"quoted text with whitespace"
"text"
"text with whitespace"

$ dig +short TXT txt.shopify.test @ns1.p19.dynect.net
"quoted text with whitespace"
"text"
"text with whitespace"

$ dig +short TXT txt.shopify.test @ns-cloud-a4.googledomains.com
"text"
"text" "with" "whitespace"
"quoted text with whitespace"
```

and `record-store diff` shows...

```
$ bin/record-store diff
Diffing 2 zones
.
Zone: shopify.test
--------------------
Provider: DynECT
Update:
 - [TXTRecord] txt.shopify.test. 30 IN TXT ""quoted text with whitespace""
--------------------
Provider: GoogleCloudDNS
Update:
 - [TXTRecord] txt.shopify.test. 30 IN TXT "text with whitespace"
 - [TXTRecord] txt.shopify.test. 30 IN TXT "text"
====================
```

### PR DESCRIPTION
**Pull request by sbfaulkner**
this PR fixes #58

consistent handling of rdata for TXT records and escaping of `"` characters in strings
strip wrapping quotes from google API rdata

UPDATE: rebased after merging #55 #56 and #57

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -141,6 +119,35 @@ class Provider
...
+  def remove(record)
+  end
+
+  # Updates an existing record in the zone. It is expected this call modifies external state.
+  #
+  # Arguments:
+  # id - provider specific ID of record to update
+  # record - a kind of `Record` which the record with `id` should be updated to
+  def update(id, record)
+  end
```
Comment (albiere) [inline README.md]: This changes have been made here https://github.com/Shopify/record_store/pull/57 and do not need to be in this PR.

File: dev.yml [hunk 2]
```diff
@@ -2,7 +2,7 @@
 name: recordstore
 
 up:
-  - ruby: 2.3.1
+  - ruby: 2.4.3
```
Comment (albiere) [inline dev.yml]: Should this be in a different PR or you rely on this version to have things working properly?

File: lib/record_store/provider.rb [hunk 3]
```diff
@@ -116,6 +104,20 @@ def freezable?
...
+        raise NotImplementedError
+      end
+
+      def remove(record)
+        raise NotImplementedError
+      end
+
+      def update(id, record)
+        raise NotImplementedError
+      end
```
Comment (albiere) [inline lib/record_store/provider.rb]: Same as previous comment.

File: lib/record_store/record.rb [hunk 4]
```diff
@@ -35,7 +35,7 @@ def to_hash
     end
 
     def type
-      self.class.name.split('::').last
+      self.class.name.demodulize
```
Comment (albiere) [inline lib/record_store/record.rb]: 👍

File: lib/record_store/record/txt.rb [hunk 5]
```diff
@@ -10,16 +10,12 @@ def initialize(record)
...
-      "[#{type}Record] #{fqdn} #{ttl} IN #{type} \"#{rdata_txt}\""
-    end
-
     def rdata
       { txtdata: txtdata }
     end
 
     def rdata_txt
-      txtdata
+      %("#{txtdata.gsub('"', '\"')}")
```
Comment (albiere) [inline lib/record_store/record/txt.rb]: This logic is repeated in the `Record#build_from_api`. Do we still need this?

File: lib/record_store/provider/google_cloud_dns.rb [hunk 6]
```diff
@@ -98,7 +98,8 @@ def build_from_api(record)
         when 'NS'
           record_params.merge!(nsdname: record.data[0])
         when 'SPF', 'TXT'
-          record_params.merge!(txtdata: record.data[0].gsub(';', '\;'))
+          txtdata = record.data[0].sub(/\A"(.*)"\z/, '\1').gsub('\"', '"').gsub(';', '\;')
+          record_params.merge!(txtdata: txtdata)
```
Comment (albiere) [inline lib/record_store/provider/google_cloud_dns.rb]: Do you think it's worth moving the txtdata logic to the parent class and reuse it in the subclasses of Provider? I'm on the fence because the implementation is not quite the same, but looking at the regex, it feels like it would just work.

File: lib/record_store/provider/dynect.rb [hunk 7]
```diff
@@ -82,16 +82,28 @@ def secrets
...
+        record = api_record.merge(rdata).slice!('rdata').symbolize_keys
 
-        return if record.fetch(:record_type) == 'SOA'
+        type = record.fetch(:record_type)
+        return if type == 'SOA'
 
-        unless record.fetch(:fqdn).ends_with?('.')
-          record[:fqdn] = "#{record.fetch(:fqdn)}."
-        end
+        record[:txtdata].gsub!('\"', '"') if type == 'TXT'
```
Comment (albiere) [inline lib/record_store/provider/dynect.rb]: Just `TXT`? Do we need to also include SPF, just like we do for Google and DNSimple?

**Comment by sbfaulkner** [comment]
:tophat: ...

adding to `shopify.test.yml` (on my machine)...
```
  - type: TXT
    fqdn: txt.shopify.test.
    ttl: 30
    txtdata: text
  - type: TXT
    fqdn: txt.shopify.test.
    ttl: 30
    txtdata: "text with whitespace"
  - type: TXT
    fqdn: txt.shopify.test.
    ttl: 30
    txtdata: '"quoted text with whitespace"'
```

```
$ bin/record-store diff
Diffing 1 zones

Zone: shopify.test
--------------------
Provider: DNSimple
Add:
 - [TXTRecord] txt.shopify.test. 30 IN TXT "\"quoted text with whitespace\""
 - [TXTRecord] txt.shopify.test. 30 IN TXT "text with whitespace"
 - [TXTRecord] txt.shopify.test. 30 IN TXT "text"
--------------------
Provider: DynECT
Add:
 - [TXTRecord] txt.shopify.test. 30 IN TXT "\"quoted text with whitespace\""
 - [TXTRecord] txt.shopify.test. 30 IN TXT "text with whitespace"
 - [TXTRecord] txt.shopify.test. 30 IN TXT "text"
--------------------
Provider: GoogleCloudDNS
Add:
 - [TXTRecord] txt.shopify.test. 30 IN TXT "\"quoted text with whitespace\""
 - [TXTRecord] txt.shopify.test. 30 IN TXT "text with whitespace"
 - [TXTRecord] txt.shopify.test. 30 IN TXT "text"
====================

$ bin/record-store apply
Applying 3 additions, 0 removals, and 0 updates...
Creating [TXTRecord] txt.shopify.test. 30 IN TXT "\"quoted text with whitespace\""...
Creating [TXTRecord] txt.shopify.test. 30 IN TXT "text with whitespace"...
Creating [TXTRecord] txt.shopify.test. 30 IN TXT "text"...
Published shopify.test changes to DNSimple

Applying 3 additions, 0 removals, and 0 updates...
Creating [TXTRecord] txt.shopify.test. 30 IN TXT "\"quoted text with whitespace\""...
Creating [TXTRecord] txt.shopify.test. 30 IN TXT "text with whitespace"...
Creating [TXTRecord] txt.shopify.test. 30 IN TXT "text"...
Published shopify.test changes to DynECT

Applying 3 additions, 0 removals, and 0 updates...
Published shopify.test changes to GoogleCloudDNS

All zone changes deployed

$ bin/record-store diff
Diffing 1 zones
.

$ bin/record-store assert_empty_diff
$ echo $?
0
```

File: dev.yml (hunk 2 above)
Comment (sbfaulkner) [reply dev.yml]: I can remove... but was small enough I thought you'd let me away without doing a separate PR :-P

File: lib/record_store/record/txt.rb (hunk 5 above)
Comment (sbfaulkner) [reply lib/record_store/record/txt.rb]: `build_from_api` is unescaping
this is escaping

**Comment by sbfaulkner** [comment]
@albiere see 8e3df9a and 5919eb1

I think these address your concerns

File: dev.yml (hunk 2 above)
Comment (albiere) [reply dev.yml]: If you think that won't cause any problem because of the ruby bump, I'm fine with it 👍

**Comment by albiere** [comment]
@sbfaulkner I like that a lot 👍
